File: tools/certlang/python/check.py
#!/usr/bin/env python3
import argparse
import hashlib
import json
import os
import re
from pathlib import Path

VERSION = "PLCI/1"
EVIDENCE_TYPES = {"source", "migration", "report", "telemetry", "spec"}
OBLIGATION_KINDS = {"scope", "frame", "invariant", "rollback", "evidence", "interchange"}
STATUSES = {"checked", "assumed", "refuted", "unsupported"}
VERDICTS = {"safe", "guarded", "blocked", "unsupported"}
ID_RE = re.compile(r"^[a-z][a-z0-9.-]{2,80}$")
ISSUED_RE = re.compile(r"^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z$")
REPO_RE = re.compile(r"^[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+$")
REF_RE = re.compile(r"^[A-Za-z0-9._/-]{1,80}$")


def valid_id(value):
    return bool(ID_RE.match(value))


def valid_sha256(value):
    return len(value) == 64 and all(ch in "0123456789abcdef" for ch in value)


def valid_local_path(value):
    if not value or value.startswith("/") or "\\" in value:
        return False
    parts = value.split("/")
    return all(part and part not in {".", ".."} and re.match(r"^[A-Za-z0-9_.-]+$", part) for part in parts)


def sha256_bytes(data):
    return hashlib.sha256(data).hexdigest()


def parse_field(lines, index, prefix):
    if index >= len(lines) or not lines[index].startswith(prefix):
        raise ValueError(f"line {index + 1}: expected {prefix[:-2]} field")
    value = lines[index][len(prefix):]
    if not value:
        raise ValueError(f"line {index + 1}: {prefix[:-2]} must not be empty")
    return value, index + 1


def parse_evidence(line, index):
    parts = line[len("evidence: "):].split(" ")
    if len(parts) != 4 or any(part == "" for part in parts):
        raise ValueError(f"line {index + 1}: evidence line must have four single-spaced fields")
    if not parts[1].startswith("type=") or not parts[2].startswith("uri=") or not parts[3].startswith("sha256="):
        raise ValueError(f"line {index + 1}: evidence attributes are out of order")
    return {
        "id": parts[0],
        "type": parts[1][5:],
        "uri": parts[2][4:],
        "sha256": parts[3][7:],
    }


def parse_obligation(line, index):
    rest = line[len("obligation: "):]
    marker = ' formula="'
    if marker not in rest or not rest.endswith('"'):
        raise ValueError(f"line {index + 1}: obligation must end with formula=\"...\"")
    before, formula = rest.split(marker, 1)
    formula = formula[:-1]
    parts = before.split(" ")
    if len(parts) != 4 or any(part == "" for part in parts):
        raise ValueError(f"line {index + 1}: obligation line must have five single-spaced fields")
    if not parts[1].startswith("kind=") or not parts[2].startswith("status=") or not parts[3].startswith("evidence="):
        raise ValueError(f"line {index + 1}: obligation attributes are out of order")
    return {
        "id": parts[0],
        "kind": parts[1][5:],
        "status": parts[2][7:],
        "evidence": parts[3][9:].split(","),
        "formula": formula,
    }


def verify_file(root, path, expected):
    if not valid_local_path(path):
        raise ValueError(f"invalid local path {path!r}")
    data = (root / Path(path)).read_bytes()
    got = sha256_bytes(data)
    if got != expected:
        raise ValueError(f"{path} sha256 got {got} want {expected}")


def parse_certificate(path, root):
    raw = path.read_bytes()
    if not raw:
        raise ValueError("empty certificate")
    if b"\r" in raw:
        raise ValueError("PLCI certificates must use LF line endings")
    if not raw.endswith(b"\n"):
        raise ValueError("PLCI certificates must end with LF")
    lines = raw.decode("utf-8").rstrip("\n").split("\n")
    if len(lines) < 14:
        raise ValueError("certificate is shorter than the PLCI/1 grammar minimum")
    index = 0
    if lines[index] != VERSION:
        raise ValueError("line 1: expected PLCI/1")
    index += 1
    cert = {}
    for key, prefix in [
        ("certificate_id", "certificate-id: "),
        ("subject_repo", "subject-repo: "),
        ("subject_ref", "subject-ref: "),
        ("subject_path", "subject-path: "),
        ("subject_sha256", "subject-sha256: "),
        ("issued_at", "issued-at: "),
        ("producer", "producer: "),
        ("verdict", "verdict: "),
        ("risk_bps", "risk-bps: "),
    ]:
        cert[key], index = parse_field(lines, index, prefix)
    if not re.match(r"^(0|10000|[1-9][0-9]{0,3})$", cert["risk_bps"]):
        raise ValueError("risk-bps must be 0..10000 without leading zeros")
    cert["risk_bps"] = int(cert["risk_bps"])

    evidence = []
    while index < len(lines) and lines[index].startswith("evidence: "):
        evidence.append(parse_evidence(lines[index], index))
        index += 1
    if not evidence:
        raise ValueError(f"line {index + 1}: expected at least one evidence line")
    obligations = []
    while index < len(lines) and lines[index].startswith("obligation: "):
        obligations.append(parse_obligation(lines[index], index))
        index += 1
    if not obligations:
        raise ValueError(f"line {index + 1}: expected at least one obligation line")
    canonical_index = index
    cert["canonical_sha256"], index = parse_field(lines, index, "canonical-sha256: ")
    if index >= len(lines) or lines[index] != "END":
        raise ValueError(f"line {index + 1}: expected END")
    index += 1
    if index != len(lines):
        raise ValueError(f"line {index + 1}: unexpected trailing line")

    canonical = ("\n".join(lines[:canonical_index]) + "\n").encode("utf-8")
    got = sha256_bytes(canonical)
    if cert["canonical_sha256"] != got:
        raise ValueError(f"canonical-sha256 mismatch: got {cert['canonical_sha256']} want {got}")

    validate_certificate(cert, evidence, obligations, root)


def validate_certificate(cert, evidence, obligations, root):
    if not valid_id(cert["certificate_id"]):
        raise ValueError(f"invalid certificate-id {cert['certificate_id']!r}")
    if not REPO_RE.match(cert["subject_repo"]):
        raise ValueError(f"invalid subject-repo {cert['subject_repo']!r}")
    if not REF_RE.match(cert["subject_ref"]):
        raise ValueError(f"invalid subject-ref {cert['subject_ref']!r}")
    if not valid_local_path(cert["subject_path"]):
        raise ValueError(f"invalid subject-path {cert['subject_path']!r}")
    if not valid_sha256(cert["subject_sha256"]):
        raise ValueError(f"invalid subject-sha256 {cert['subject_sha256']!r}")
    if not valid_issued_at(cert["issued_at"]):
        raise ValueError(f"invalid issued-at {cert['issued_at']!r}")
    if not valid_vchar(cert["producer"]):
        raise ValueError(f"invalid producer {cert['producer']!r}")
    if cert["verdict"] not in VERDICTS:
        raise ValueError(f"invalid verdict {cert['verdict']!r}")
    if cert["risk_bps"] < 0 or cert["risk_bps"] > 10000:
        raise ValueError(f"risk-bps out of range: {cert['risk_bps']}")
    verify_file(root, cert["subject_path"], cert["subject_sha256"])

    evidence_ids = set()
    for item in evidence:
        if not valid_id(item["id"]):
            raise ValueError(f"invalid evidence id {item['id']!r}")
        if item["id"] in evidence_ids:
            raise ValueError(f"duplicate evidence id {item['id']!r}")
        evidence_ids.add(item["id"])
        if item["type"] not in EVIDENCE_TYPES:
            raise ValueError(f"invalid evidence type {item['type']!r}")
        if not valid_sha256(item["sha256"]):
            raise ValueError(f"invalid evidence sha256 for {item['id']!r}")
        if item["uri"].startswith("file:"):
            verify_file(root, item["uri"][5:], item["sha256"])
        elif not (valid_remote_uri(item["uri"], "https://") or valid_remote_uri(item["uri"], "repo://")):
            raise ValueError(f"unsupported evidence uri {item['uri']!r}")

    status_counts = {status: 0 for status in STATUSES}
    obligation_ids = set()
    for item in obligations:
        if not valid_id(item["id"]):
            raise ValueError(f"invalid obligation id {item['id']!r}")
        if item["id"] in obligation_ids:
            raise ValueError(f"duplicate obligation id {item['id']!r}")
        obligation_ids.add(item["id"])
        if item["kind"] not in OBLIGATION_KINDS:
            raise ValueError(f"invalid obligation kind {item['kind']!r}")
        if item["status"] not in STATUSES:
            raise ValueError(f"invalid obligation status {item['status']!r}")
        if not valid_formula(item["formula"]):
            raise ValueError(f"invalid formula for obligation {item['id']!r}")
        status_counts[item["status"]] += 1
        for ref in item["evidence"]:
            if ref not in evidence_ids:
                raise ValueError(f"obligation {item['id']!r} references missing evidence {ref!r}")

    if cert["verdict"] == "safe" and any(status_counts[s] for s in ["assumed", "refuted", "unsupported"]):
        raise ValueError("safe verdict cannot carry assumed, refuted, or unsupported obligations")
    if cert["verdict"] == "guarded" and any(status_counts[s] for s in ["refuted", "unsupported"]):
        raise ValueError("guarded verdict cannot carry refuted or unsupported obligations")
    if cert["verdict"] == "blocked" and status_counts["refuted"] == 0:
        raise ValueError("blocked verdict requires at least one refuted obligation")
    if cert["verdict"] == "unsupported" and status_counts["unsupported"] == 0:
        raise ValueError("unsupported verdict requires at least one unsupported obligation")


def check_directory(spec_dir, root):
    vectors = []
    accepted = rejected = 0
    for dirname, expected in [("valid", "valid"), ("invalid", "invalid")]:
        for path in sorted((spec_dir / "vectors" / dirname).glob("*.plci")):
            error = None
            try:
                parse_certificate(path, root)
                is_accepted = True
                accepted += 1
            except Exception as exc:  # reported as checker output, not swallowed
                error = str(exc)
                is_accepted = False
                rejected += 1
            ok = (expected == "valid" and is_accepted) or (expected == "invalid" and not is_accepted)
            row = {
                "path": str(path.relative_to(spec_dir / "vectors")).replace(os.sep, "/"),
                "expected": expected,
                "accepted": is_accepted,
                "ok": ok,
            }
            if error:
                row["error"] = error
            vectors.append(row)
    return {
        "checker": "python",
        "version": VERSION,
        "spec_dir": str(spec_dir),
        "total_valid": sum(1 for row in vectors if row["expected"] == "valid"),
        "total_invalid": sum(1 for row in vectors if row["expected"] == "invalid"),
        "accepted": accepted,
        "rejected": rejected,
        "all_ok": bool(vectors) and all(row["ok"] for row in vectors),
        "vectors": vectors,
    }


def valid_issued_at(value):
    if not ISSUED_RE.match(value):
        return False
    year = int(value[0:4])
    month = int(value[5:7])
    day = int(value[8:10])
    hour = int(value[11:13])
    minute = int(value[14:16])
    second = int(value[17:19])
    if month < 1 or month > 12 or hour > 23 or minute > 59 or second > 59:
        return False
    days = [31, 29 if leap_year(year) else 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    return 1 <= day <= days[month - 1]


def leap_year(year):
    return year % 400 == 0 or (year % 4 == 0 and year % 100 != 0)


def valid_remote_uri(value, scheme):
    if not value.startswith(scheme):
        return False
    rest = value[len(scheme):]
    return bool(rest) and all(0x21 <= ord(ch) <= 0x7E for ch in rest)


def valid_vchar(value):
    return bool(value) and all(0x21 <= ord(ch) <= 0x7E for ch in value)


def valid_formula(value):
    return bool(value) and all(ch != '"' and 0x20 <= ord(ch) <= 0x7E for ch in value)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--spec-dir", default="specs/certificate-interchange/v1")
    parser.add_argument("--root", default=".")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()
    report = check_directory(Path(args.spec_dir), Path(args.root))
    if args.json:
        print(json.dumps(report, indent=2))
    else:
        print(f"python PLCI/1 checker: valid={report['total_valid']} invalid={report['total_invalid']} all_ok={report['all_ok']}")
    raise SystemExit(0 if report["all_ok"] else 1)


if __name__ == "__main__":
    main()

File: tools/certlang/typescript/check.ts
import * as crypto from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";

const VERSION = "PLCI/1";
const EVIDENCE_TYPES = new Set(["source", "migration", "report", "telemetry", "spec"]);
const OBLIGATION_KINDS = new Set(["scope", "frame", "invariant", "rollback", "evidence", "interchange"]);
const STATUSES = new Set(["checked", "assumed", "refuted", "unsupported"]);
const VERDICTS = new Set(["safe", "guarded", "blocked", "unsupported"]);

type Evidence = { id: string; type: string; uri: string; sha256: string };
type Obligation = { id: string; kind: string; status: string; evidence: string[]; formula: string };
type Cert = {
  certificateId: string;
  subjectRepo: string;
  subjectRef: string;
  subjectPath: string;
  subjectSha256: string;
  issuedAt: string;
  producer: string;
  verdict: string;
  riskBps: number;
  canonicalSha256: string;
  evidence: Evidence[];
  obligations: Obligation[];
};

function validID(value: string): boolean {
  return /^[a-z][a-z0-9.-]{2,80}$/.test(value);
}

function validSHA256(value: string): boolean {
  return /^[0-9a-f]{64}$/.test(value);
}

function validLocalPath(value: string): boolean {
  return value.length > 0 && !value.startsWith("/") && !value.includes("\\") && value.split("/").every((part) => Boolean(part) && part !== "." && part !== ".." && /^[A-Za-z0-9_.-]+$/.test(part));
}

function sha256(data: Buffer | string): string {
  return crypto.createHash("sha256").update(data).digest("hex");
}

function lineError(index: number, message: string): Error {
  return new Error(`line ${index + 1}: ${message}`);
}

function parseField(lines: string[], state: { index: number }, prefix: string): string {
  if (state.index >= lines.length || !lines[state.index].startsWith(prefix)) {
    throw lineError(state.index, `expected ${prefix.slice(0, -2)} field`);
  }
  const value = lines[state.index].slice(prefix.length);
  if (!value) {
    throw lineError(state.index, `${prefix.slice(0, -2)} must not be empty`);
  }
  state.index += 1;
  return value;
}

function parseEvidenceLine(line: string, index: number): Evidence {
  const parts = line.slice("evidence: ".length).split(" ");
  if (parts.length !== 4 || parts.some((part) => part === "")) {
    throw lineError(index, "evidence line must have four single-spaced fields");
  }
  if (!parts[1].startsWith("type=") || !parts[2].startsWith("uri=") || !parts[3].startsWith("sha256=")) {
    throw lineError(index, "evidence attributes are out of order");
  }
  return {
    id: parts[0],
    type: parts[1].slice("type=".length),
    uri: parts[2].slice("uri=".length),
    sha256: parts[3].slice("sha256=".length),
  };
}

function parseObligationLine(line: string, index: number): Obligation {
  const rest = line.slice("obligation: ".length);
  const marker = ' formula="';
  const markerIndex = rest.indexOf(marker);
  if (markerIndex < 0 || !rest.endsWith('"')) {
    throw lineError(index, 'obligation must end with formula="..."');
  }
  const before = rest.slice(0, markerIndex);
  const formula = rest.slice(markerIndex + marker.length, -1);
  const parts = before.split(" ");
  if (parts.length !== 4 || parts.some((part) => part === "")) {
    throw lineError(index, "obligation line must have five single-spaced fields");
  }
  if (!parts[1].startsWith("kind=") || !parts[2].startsWith("status=") || !parts[3].startsWith("evidence=")) {
    throw lineError(index, "obligation attributes are out of order");
  }
  return {
    id: parts[0],
    kind: parts[1].slice("kind=".length),
    status: parts[2].slice("status=".length),
    evidence: parts[3].slice("evidence=".length).split(","),
    formula,
  };
}

function verifyFile(root: string, value: string, expected: string): void {
  if (!validLocalPath(value)) {
    throw new Error(`invalid local path ${JSON.stringify(value)}`);
  }
  const fullPath = path.join(root, ...value.split("/"));
  const got = sha256(fs.readFileSync(fullPath));
  if (got !== expected) {
    throw new Error(`${value} sha256 got ${got} want ${expected}`);
  }
}

function parseCertificate(filePath: string, root: string): void {
  const raw = fs.readFileSync(filePath);
  if (raw.length === 0) {
    throw new Error("empty certificate");
  }
  const rawText = raw.toString("utf8");
  if (rawText.includes("\r")) {
    throw new Error("PLCI certificates must use LF line endings");
  }
  if (!rawText.endsWith("\n")) {
    throw new Error("PLCI certificates must end with LF");
  }
  const lines = rawText.slice(0, -1).split("\n");
  if (lines.length < 14) {
    throw new Error("certificate is shorter than the PLCI/1 grammar minimum");
  }
  const state = { index: 0 };
  if (lines[state.index] !== VERSION) {
    throw lineError(state.index, "expected PLCI/1");
  }
  state.index += 1;
  const certificateId = parseField(lines, state, "certificate-id: ");
  const subjectRepo = parseField(lines, state, "subject-repo: ");
  const subjectRef = parseField(lines, state, "subject-ref: ");
  const subjectPath = parseField(lines, state, "subject-path: ");
  const subjectSha256 = parseField(lines, state, "subject-sha256: ");
  const issuedAt = parseField(lines, state, "issued-at: ");
  const producer = parseField(lines, state, "producer: ");
  const verdict = parseField(lines, state, "verdict: ");
  const riskBpsText = parseField(lines, state, "risk-bps: ");
  if (!/^(0|10000|[1-9][0-9]{0,3})$/.test(riskBpsText)) {
    throw new Error("risk-bps must be 0..10000 without leading zeros");
  }
  const riskBps = Number(riskBpsText);
  const evidence: Evidence[] = [];
  while (state.index < lines.length && lines[state.index].startsWith("evidence: ")) {
    evidence.push(parseEvidenceLine(lines[state.index], state.index));
    state.index += 1;
  }
  if (evidence.length === 0) {
    throw lineError(state.index, "expected at least one evidence line");
  }
  const obligations: Obligation[] = [];
  while (state.index < lines.length && lines[state.index].startsWith("obligation: ")) {
    obligations.push(parseObligationLine(lines[state.index], state.index));
    state.index += 1;
  }
  if (obligations.length === 0) {
    throw lineError(state.index, "expected at least one obligation line");
  }
  const canonicalIndex = state.index;
  const canonicalSha256 = parseField(lines, state, "canonical-sha256: ");
  if (state.index >= lines.length || lines[state.index] !== "END") {
    throw lineError(state.index, "expected END");
  }
  state.index += 1;
  if (state.index !== lines.length) {
    throw lineError(state.index, "unexpected trailing line");
  }
  const got = sha256(`${lines.slice(0, canonicalIndex).join("\n")}\n`);
  if (got !== canonicalSha256) {
    throw new Error(`canonical-sha256 mismatch: got ${canonicalSha256} want ${got}`);
  }
  validateCertificate({
    certificateId,
    subjectRepo,
    subjectRef,
    subjectPath,
    subjectSha256,
    issuedAt,
    producer,
    verdict,
    riskBps,
    canonicalSha256,
    evidence,
    obligations,
  }, root);
}

function validateCertificate(cert: Cert, root: string): void {
  if (!validID(cert.certificateId)) throw new Error(`invalid certificate-id ${JSON.stringify(cert.certificateId)}`);
  if (!/^[A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+$/.test(cert.subjectRepo)) throw new Error(`invalid subject-repo ${JSON.stringify(cert.subjectRepo)}`);
  if (!/^[A-Za-z0-9._/-]{1,80}$/.test(cert.subjectRef)) throw new Error(`invalid subject-ref ${JSON.stringify(cert.subjectRef)}`);
  if (!validLocalPath(cert.subjectPath)) throw new Error(`invalid subject-path ${JSON.stringify(cert.subjectPath)}`);
  if (!validSHA256(cert.subjectSha256) || !validSHA256(cert.canonicalSha256)) throw new Error("invalid sha256 field");
  if (!validIssuedAt(cert.issuedAt)) throw new Error(`invalid issued-at ${JSON.stringify(cert.issuedAt)}`);
  if (!validVChar(cert.producer)) throw new Error(`invalid producer ${JSON.stringify(cert.producer)}`);
  if (!VERDICTS.has(cert.verdict)) throw new Error(`invalid verdict ${JSON.stringify(cert.verdict)}`);
  if (cert.riskBps < 0 || cert.riskBps > 10000) throw new Error(`risk-bps out of range: ${cert.riskBps}`);
  verifyFile(root, cert.subjectPath, cert.subjectSha256);

  const evidenceIDs = new Set<string>();
  for (const item of cert.evidence) {
    if (!validID(item.id)) throw new Error(`invalid evidence id ${JSON.stringify(item.id)}`);
    if (evidenceIDs.has(item.id)) throw new Error(`duplicate evidence id ${JSON.stringify(item.id)}`);
    evidenceIDs.add(item.id);
    if (!EVIDENCE_TYPES.has(item.type)) throw new Error(`invalid evidence type ${JSON.stringify(item.type)}`);
    if (!validSHA256(item.sha256)) throw new Error(`invalid evidence sha256 for ${JSON.stringify(item.id)}`);
    if (item.uri.startsWith("file:")) {
      verifyFile(root, item.uri.slice("file:".length), item.sha256);
    } else if (!validRemoteURI(item.uri, "https://") && !validRemoteURI(item.uri, "repo://")) {
      throw new Error(`unsupported evidence uri ${JSON.stringify(item.uri)}`);
    }
  }

  const obligationIDs = new Set<string>();
  const statusCounts = new Map<string, number>();
  for (const item of cert.obligations) {
    if (!validID(item.id)) throw new Error(`invalid obligation id ${JSON.stringify(item.id)}`);
    if (obligationIDs.has(item.id)) throw new Error(`duplicate obligation id ${JSON.stringify(item.id)}`);
    obligationIDs.add(item.id);
    if (!OBLIGATION_KINDS.has(item.kind)) throw new Error(`invalid obligation kind ${JSON.stringify(item.kind)}`);
    if (!STATUSES.has(item.status)) throw new Error(`invalid obligation status ${JSON.stringify(item.status)}`);
    if (!validFormula(item.formula)) throw new Error(`invalid formula for obligation ${JSON.stringify(item.id)}`);
    statusCounts.set(item.status, (statusCounts.get(item.status) ?? 0) + 1);
    for (const ref of item.evidence) {
      if (!evidenceIDs.has(ref)) throw new Error(`obligation ${JSON.stringify(item.id)} references missing evidence ${JSON.stringify(ref)}`);
    }
  }
  const count = (status: string) => statusCounts.get(status) ?? 0;
  if (cert.verdict === "safe" && (count("assumed") > 0 || count("refuted") > 0 || count("unsupported") > 0)) {
    throw new Error("safe verdict cannot carry assumed, refuted, or unsupported obligations");
  }
  if (cert.verdict === "guarded" && (count("refuted") > 0 || count("unsupported") > 0)) {
    throw new Error("guarded verdict cannot carry refuted or unsupported obligations");
  }
  if (cert.verdict === "blocked" && count("refuted") === 0) {
    throw new Error("blocked verdict requires at least one refuted obligation");
  }
  if (cert.verdict === "unsupported" && count("unsupported") === 0) {
    throw new Error("unsupported verdict requires at least one unsupported obligation");
  }
}

function validIssuedAt(value: string): boolean {
  if (!/^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z$/.test(value)) return false;
  const year = Number(value.slice(0, 4));
  const month = Number(value.slice(5, 7));
  const day = Number(value.slice(8, 10));
  const hour = Number(value.slice(11, 13));
  const minute = Number(value.slice(14, 16));
  const second = Number(value.slice(17, 19));
  if (month < 1 || month > 12 || hour > 23 || minute > 59 || second > 59) return false;
  const days = [31, leapYear(year) ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  return day >= 1 && day <= days[month - 1];
}

function leapYear(year: number): boolean {
  return year % 400 === 0 || (year % 4 === 0 && year % 100 !== 0);
}

function validRemoteURI(value: string, scheme: string): boolean {
  if (!value.startsWith(scheme)) return false;
  const rest = value.slice(scheme.length);
  return rest.length > 0 && /^[\x21-\x7E]+$/.test(rest);
}

function validVChar(value: string): boolean {
  return value.length > 0 && /^[\x21-\x7E]+$/.test(value);
}

function validFormula(value: string): boolean {
  return value.length > 0 && /^[\x20-\x21\x23-\x7E]+$/.test(value);
}

function checkDirectory(specDir: string, root: string) {
  const vectors = [];
  let accepted = 0;
  let rejected = 0;
  for (const [dirname, expected] of [["valid", "valid"], ["invalid", "invalid"]] as const) {
    const dir = path.join(specDir, "vectors", dirname);
    const files = fs.readdirSync(dir).filter((name) => name.endsWith(".plci")).sort();
    for (const file of files) {
      const filePath = path.join(dir, file);
      let isAccepted = false;
      let error: string | undefined;
      try {
        parseCertificate(filePath, root);
        isAccepted = true;
        accepted += 1;
      } catch (err) {
        error = err instanceof Error ? err.message : String(err);
        rejected += 1;
      }
      const ok = (expected === "valid" && isAccepted) || (expected === "invalid" && !isAccepted);
      const row: Record<string, string | boolean> = {
        path: path.relative(path.join(specDir, "vectors"), filePath).replaceAll(path.sep, "/"),
        expected,
        accepted: isAccepted,
        ok,
      };
      if (error) row.error = error;
      vectors.push(row);
    }
  }
  return {
    checker: "typescript",
    version: VERSION,
    spec_dir: specDir,
    total_valid: vectors.filter((row) => row.expected === "valid").length,
    total_invalid: vectors.filter((row) => row.expected === "invalid").length,
    accepted,
    rejected,
    all_ok: vectors.length > 0 && vectors.every((row) => row.ok === true),
    vectors,
  };
}

function main(): void {
  const args = process.argv.slice(2);
  let specDir = "specs/certificate-interchange/v1";
  let root = ".";
  let json = false;
  for (let i = 0; i < args.length; i += 1) {
    if (args[i] === "--spec-dir") specDir = args[++i];
    else if (args[i] === "--root") root = args[++i];
    else if (args[i] === "--json") json = true;
    else throw new Error(`unknown argument ${args[i]}`);
  }
  const report = checkDirectory(specDir, root);
  if (json) {
    console.log(JSON.stringify(report, null, 2));
  } else {
    console.log(`typescript PLCI/1 checker: valid=${report.total_valid} invalid=${report.total_invalid} all_ok=${report.all_ok}`);
  }
  process.exit(report.all_ok ? 0 : 1);
}

main();
